### PR TITLE
[GKD] add liger loss

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Added Liger GKD loss and test

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/gkd_trainer.py [hunk 1]
```diff
@@ -77,9 +80,29 @@ def __init__(
...
+        # Ensure SFTTrainer does not pre-process the dataset when using a ChatML collator,
+        # so that raw conversational fields (e.g., "messages") remain available to the collator.
+        if getattr(args, "dataset_kwargs", None) is None:
+            args.dataset_kwargs = {"skip_prepare_dataset": True}
+        else:
+            args.dataset_kwargs["skip_prepare_dataset"] = True
+
+        # Liger fused GKD loss (JSD)
+        self.use_liger_gkd_loss = False
+        if getattr(args, "use_liger_kernel", False):
```
Comment (qgallouedec) [inline trl/trainer/gkd_trainer.py]:
```suggestion
        if args.use_liger_kernel:
```

File: trl/trainer/gkd_trainer.py [hunk 2]
```diff
@@ -77,9 +80,29 @@ def __init__(
...
+        # Ensure Trainer does not drop non-signature columns used by the collator (e.g., "prompts")
         args.remove_unused_columns = False
-        data_collator = DataCollatorForChatML(tokenizer=processing_class, max_length=args.max_length)
+        # Respect a user-provided data_collator; otherwise, provide a ChatML collator that
+        if data_collator is None:
+            data_collator = DataCollatorForChatML(tokenizer=processing_class, max_length=args.max_length)
+
+        # Ensure SFTTrainer does not pre-process the dataset when using a ChatML collator,
+        # so that raw conversational fields (e.g., "messages") remain available to the collator.
+        if getattr(args, "dataset_kwargs", None) is None:
```
Comment (qgallouedec) [inline trl/trainer/gkd_trainer.py]:
```suggestion
        if args.dataset_kwargs is None:
```

File: tests/test_gkd_trainer.py [hunk 3]
```diff
@@ -244,6 +245,36 @@ def test_gkd_trainer(self):
...
         self.assertIn("model.safetensors", os.listdir(self.tmp_dir + "/checkpoint-2"))
 
+    @require_liger_kernel
+    def test_gkd_trainer_with_liger(self):
+        training_args = GKDConfig(
+            output_dir=self.tmp_dir,
+            dataloader_drop_last=True,
+            eval_strategy="no",
+            max_steps=2,
+            per_device_train_batch_size=2,
```
Comment (qgallouedec) [inline tests/test_gkd_trainer.py]:
```suggestion
```

File: trl/trainer/gkd_trainer.py [hunk 4]
```diff
@@ -217,33 +240,92 @@ def generalized_jsd_loss(
...
+                student_weight=student_head.weight,
+                teacher_input=teacher_hidden,
+                teacher_weight=teacher_head.weight,
+                true_labels=true_labels,
+                student_bias=getattr(student_head, "bias", None),
+                teacher_bias=getattr(teacher_head, "bias", None),
+            )
+        else:
+            # compute student output
+            outputs_student = model(
```
Comment (qgallouedec) [inline trl/trainer/gkd_trainer.py]: Let's try to be consistent in the naming: either `student_output` or `output_student` (same for teacher)

**Review by qgallouedec — APPROVED** [review]
Nice! Just a few nits 
Does it require to bump liger?

**Comment by kashif** [comment]
its supported from version 0.5.7 onwards
